src/python_framework_v02/troute/nhd_network_augment.py

```diff
@@ -159,49 +167,138 @@ def build_reaches(rconn):
     for tw, net in subnets.items():
         path_func = partial(nhd_network.split_at_junction, net)
         subreaches[tw] = nhd_network.dfs_decomposition(net, path_func)
+    
+    return subreachable, subreaches
 
-    return subreachable, subreaches, subnets
+def find_headwater_reaches(subreaches, hws):
+    
+    """
+    Construct a list of headwater reaches
+    Args:
+        subreaches (dict): reaches (linearly connected segments) connected to each network tailwater (keys)
+        hws (list): headwater segments
+    Returns:
+        hw_reaches (list): lists of headwater reaches 
+    """
+    
+    hw_reaches = []
+    for sublists in list(subreaches.values()):
+        for rch in sublists:
+            for val in rch:
```

Comment:
You are essentially doing a set intersection here. It might be faster to use sets than do linear searches.
```
...
for rch in sublists:
    rch = set(rch)
    hw_reaches.extend(rch & hws)
```

Reply:
Creating the set is O(n) (and happens once), but lookups are O(1) so looking up m values will be O(n+m).
Iterating thru rch is O(n), but for each value we are executing a linear search of m items (hws) => O(n*m)

Reply:
Actually, make hws the set.
```
def find_headwater_reaches(...):
    hws = set(hws)
    for sublists in subreaches.values():
        for rch in sublists:
            hw_reaches.extend(hws & rch)
    return hw_reaches
```

Reply:
hmm. This suggestion seems to make `hw_reaches` just a list of headwater segments. Instead, we'd want `hw_reaches` to be a list of lists, where each list is a combination of connected segments comprising a reach, or a set of tuples, where each tuple is a collection of connected segments comprising a reach. 

Also, `rch` is a list so the `&` operation fails, unless we convert to a set.

Reply:
Use the set.intersection method. It accepts list. Change extend to append.
We know that headwaters can only be the first element of the reach list. Can we just check that element rather than looping thru the reach?